### PR TITLE
[serve] Implement timeouts for streaming & enable all tests with experimental streaming flag turned on

### LINKED ISSUE #35779: [serve][streaming] Implement timeout logic when using streaming interface

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

This is the penultimate PR for turning `RAY_SERVE_EXPERIMENTAL_STREAMING` on by default.

In this change, we are now running all Serve tests under three conditions:

- `RAY_SERVE_EXPERIMENTAL_STREAMING=0 && RAY_SERVE_ENABLE_NEW_ROUTING=0`
- `RAY_SERVE_EXPERIMENTAL_STREAMING=0 && RAY_SERVE_ENABLE_NEW_ROUTING=1`
- `RAY_SERVE_EXPERIMENTAL_STREAMING=1 && RAY_SERVE_ENABLE_NEW_ROUTING=1`

(Note that `RAY_SERVE_EXPERIMENTAL_STREAMING=1` implies `RAY_SERVE_ENABLE_NEW_ROUTING` so the latter is not explicitly set).

This required making two functionality fixes:

- Never using the streaming path for Java even if the flag is on.
- Implementing timeouts in the streaming path.

I could split these into separate PRs, but they can't really be fully tested independently, so opting to combine them here.

## Related issue number

Closes https://github.com/ray-project/ray/issues/35779

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/http_proxy.py [hunk 1]
```diff
@@ -650,6 +668,79 @@ async def proxy_asgi_receive(
...
+        status_code = ""
+        start = time.time()
+        while True:
+            try:
+                if timeout_s is None:
+                    curr_timeout_s = None
+                else:
+                    curr_timeout_s = timeout_s - (time.time() - start)
+
+                obj_ref = await obj_ref_generator._next_async(timeout_s=curr_timeout_s)
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/http_proxy.py:724]: What's `obj_ref_generator`'s behavior if the timeout is negative? What if it's zero?

File: python/ray/serve/_private/http_proxy.py [hunk 2]
```diff
@@ -668,22 +759,40 @@ async def send_request_to_replica_streaming(
         )
 
         status_code = ""
+        start = time.time()
+        timeout_s = RAY_SERVE_REQUEST_PROCESSING_TIMEOUT_S
```
Comment (GeneDer) [inline python/ray/serve/_private/http_proxy.py:763]: We should probably use `self.request_timeout_s` for this so the config file can take precedent, but let me know if it's intentional.

**Review by shrekris-anyscale — APPROVED** [review]
Nice, this change looks good to me overall!

File: python/ray/serve/_private/http_proxy.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/_private/http_proxy.py:724]: Good question, let me check and ensure we're being defensive here.

**Review by GeneDer — COMMENTED** [review]
Just a small comment on how we set timeout, but otherwise LGTM

File: python/ray/serve/_private/http_proxy.py (hunk 2 above)
Comment (edoakes) [reply python/ray/serve/_private/http_proxy.py:763]: oh! totally, let me fix that.

Comment (edoakes) [reply python/ray/serve/_private/http_proxy.py]: done

File: python/ray/serve/_private/http_proxy.py [hunk 3]
```diff
@@ -650,6 +671,79 @@ async def proxy_asgi_receive(
...
+            return_when=FIRST_COMPLETED,
+            timeout=timeout_s,
+        )
+        if disconnected_task in done:
+            assignment_task.cancel()
+            return None
+        elif assignment_task in done:
+            return assignment_task.result()
+        else:
+            raise TimeoutError()
```
Comment (sihanwang41) [inline python/ray/serve/_private/http_proxy.py:701]: I think you have to kill "pending" tasks (which is "_" now). Looks like a leak to me.

File: python/ray/serve/_private/http_proxy.py [hunk 4]
```diff
@@ -56,6 +58,7 @@
     "RAY_SERVE_HTTP_REQUEST_MAX_RETRIES cannot be negative."
 )
 
+TIMEOUT_ERROR_CODE = "timeout"
```
Comment (sihanwang41) [inline python/ray/serve/_private/http_proxy.py:61]: normally, the error code normally is a string of number? Is the timeout error code exposing to the users?

**Review by sihanwang41 — APPROVED** [review]
thanks for addressing my comments!

**Review by GeneDer — APPROVED** [review]
:shipit:

File: python/ray/serve/_private/http_proxy.py (hunk 1 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/http_proxy.py:728]: Couldn't `curr_timeout_s = timeout_s - (time.time() - start)` make `curr_timeout_s` negative?

File: python/ray/serve/BUILD [hunk 5]
```diff
@@ -612,7 +595,7 @@ py_test(
 
 py_test(
     name = "test_gradio",
-    size = "small",
+    size = "medium",
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/BUILD:606]: I have seen timeouts in the past, I don't see a significant different when running w/ or w/o it

**Comment by edoakes** [comment]
FYI to reviewers: the diff grew because I refactored the tests to pull the timeout tests into `test_request_timeout` and add a test for timing out while pending assignment.

File: python/ray/serve/_private/http_proxy.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/_private/http_proxy.py:728]: let me fix and add unit tests

Comment (edoakes) [reply python/ray/serve/_private/http_proxy.py]: good catch. I added this to `utils.py::calculate_remaining_timeout` and added a test case for it.

Comment (shrekris-anyscale) [reply python/ray/serve/_private/http_proxy.py]: Nice, thanks!

**Pull request merged**